packages/compose/src/hooks/use-state-with-history/index.ts

```diff
@@ -11,44 +67,28 @@ import { useCallback, useRef, useState, useEffect } from '@wordpress/element';
  * @return Value, setValue, hasUndo, hasRedo, undo, redo.
  */
 export default function useStateWithHistory< T >( initialValue: T ) {
-	const manager = useRef( createUndoManager() );
-	const [ value, setValue ] = useState( initialValue );
-	const currentValue = useRef( value );
-	useEffect( () => {
-		currentValue.current = value;
-	}, [ value ] );
+	const undoManager = useMemo( () => createUndoManager(), [] );
+	const [ state, dispatch ] = useReducer( undoRedoReducer, {
+		manager: undoManager,
+		value: initialValue,
+	} );
```

Comment:
The `useMemo` call could be merged into `useReducer`:
```js
const undoRedoInit = ( value ) => ( {
  manager: createUndoManager(),
  value,
} );

useReducer( undoRedoReducer, initialValue, undoRedoInit );
```

Reply:
What's "initialValue" in this example?

Reply:
oh got it now. I guess it feels a bit too confusing to me but I understand that it's the purpose of this argument.

Reply:
It's the `initialValue` parameter passed to the `useStateWithHistoryHook`. The `useReducer` API has two parameters for initialization. One is the value, the second is a function that converts the value to initial state. One benefit of this slightly complex setup is that the init function can be extracted outside the hook as a constant. Just like the reducer function.